atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/WriteBatchingTransactionService.java

```diff
@@ -0,0 +1,158 @@
+import java.util.stream.Collectors;
+
+import javax.annotation.CheckForNull;
+
+import org.immutables.value.Value;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.SettableFuture;
+import com.palantir.atlasdb.autobatch.BatchElement;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.atlasdb.keyvalue.api.KeyAlreadyExistsException;
+import com.palantir.common.annotation.Output;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+
+/**
+ * This class coalesces write (that is, put-unless-exists) requests to an underlying {@link TransactionService}, such
+ * that there is at most one request in flight at a given time. Read requests (gets) are not batched.
+ */
+public final class WriteBatchingTransactionService implements TransactionService, AutoCloseable {
+    private final EncodingTransactionService delegate;
+    private final DisruptorAutobatcher<TimestampPair, Void> autobatcher;
+
+    private WriteBatchingTransactionService(
+            EncodingTransactionService delegate, DisruptorAutobatcher<TimestampPair, Void> autobatcher) {
+        this.delegate = delegate;
+        this.autobatcher = autobatcher;
+    }
+
+    public static TransactionService create(EncodingTransactionService delegate) {
+        DisruptorAutobatcher<TimestampPair, Void> autobatcher = DisruptorAutobatcher.create(
+                elements -> processBatch(delegate, elements));
+        return new WriteBatchingTransactionService(delegate, autobatcher);
+    }
+
+    @CheckForNull
+    @Override
+    public Long get(long startTimestamp) {
+        return delegate.get(startTimestamp);
+    }
+
+    @Override
+    public Map<Long, Long> get(Iterable<Long> startTimestamps) {
+        return delegate.get(startTimestamps);
+    }
+
+    @Override
+    public void putUnlessExists(long startTimestamp, long commitTimestamp) throws KeyAlreadyExistsException {
+        try {
+            autobatcher.apply(TimestampPair.of(startTimestamp, commitTimestamp)).get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() {
+        autobatcher.close();
+    }
+
+    /**
+     * Semantics for batch processing:
+     *
+     * - If there are multiple requests to {@link TransactionService#putUnlessExists(long, long)} with the same
+     *   start timestamp, only the first of these will actually be attempted; the remainder will fail with a
+     *   {@link SafeIllegalArgumentException}, which is fine (following the contract it means the put may or may not
+     *   have happened).
+     * - If a {@link KeyAlreadyExistsException} is thrown, we fail out requests for keys present in the
```

Comment:
I'd give a bit more detail on why we're retrying here, and possibly why we did not opt for alternative approaches